File: Gdx/fell/core/src/com/dugsolutions/fell/FellDungeon.java
package com.dugsolutions.fell;

import com.badlogic.gdx.ApplicationAdapter;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.InputAdapter;
import com.badlogic.gdx.graphics.Camera;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.PerspectiveCamera;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureAtlas.AtlasRegion;
import com.badlogic.gdx.graphics.glutils.ShaderProgram;
import com.badlogic.gdx.math.Vector3;
import com.dugsolutions.fell.map.MapGrid;
import com.dugsolutions.fell.map.gen.GenZConstant;
import com.dugsolutions.fell.map.gen.GenZMap;

public class FellDungeon extends ApplicationAdapter {
	final static String TAG = "FellDungeon";

	static float ADJ = 0.01f;

	class MyInputAdapter extends InputAdapter {

		int triggerX0;
		int triggerX1;
		int triggerY0;
		int triggerY1;

		MyInputAdapter() {
			triggerX0 = Gdx.graphics.getWidth() / 4;
			triggerX1 = (Gdx.graphics.getWidth() * 3) / 4;
			triggerY0 = Gdx.graphics.getHeight() / 4;
			triggerY1 = (Gdx.graphics.getHeight() * 3) / 4;
		}

		boolean move(int screenX, int screenY) {
			if (screenX < triggerX0) {
				if (adjXY) {
					cam.position.x += 10;
				} else {
					cam.direction.x += ADJ;
				}
			} else if (screenX > triggerX1) {
				if (adjXY) {
					cam.position.x -= 10;
				} else {
					cam.direction.x -= ADJ;
				}
			}
			if (screenY < triggerY0) {
				if (adjXY) {
					cam.position.y -= 10;
				} else {
					cam.direction.y -= ADJ;
				}
			} else if (screenY > triggerY1) {
				if (adjXY) {
					cam.position.y += 10;
				} else {
					cam.direction.y += ADJ;
				}
			} else {
				return false;
			}
			Gdx.app.log(TAG, "CAM POS=" + cam.position.toString() + ", DIR="
					+ cam.direction.toString());
			return true;
		}

		@Override
		public boolean touchDown(int screenX, int screenY, int pointer,
				int button) {
			if (button == 1) {
				cam.direction.set(startDir);
				cam.position.set(startPos);
				return true;
			}
			return move(screenX, screenY);
		}

		@Override
		public boolean touchUp(int screenX, int screenY, int pointer, int button) {
			return false;
		}

		@Override
		public boolean touchDragged(int screenX, int screenY, int pointer) {
			return move(screenX, screenY);
		}

		@Override
		public boolean scrolled(int amount) {
			cam.position.z += 10 * amount;
			Gdx.app.log(TAG, "Z=" + cam.position.z);
			return true;
		}

		@Override
		public boolean keyTyped(char character) {
			if (character == 'z') {
				adjXY = !adjXY;
				if (adjXY) {
					Gdx.app.log(TAG, "ADJUST POSITION");
				} else {
					Gdx.app.log(TAG, "ADJUST DIRECTION");
				}
			} else if (character == 'c') {
				if (cam instanceof OrthographicCamera) {
					PerspectiveCamera pcam = new PerspectiveCamera(45,
							cam.viewportWidth, cam.viewportHeight);
					pcam.position.set(cam.position);
					pcam.direction.set(cam.direction);
					pcam.near = cam.near;
					pcam.far = cam.far;
					cam = pcam;
					Gdx.app.log(TAG, "PERSPECTIVE");

				} else {
					OrthographicCamera ocam = new OrthographicCamera(
							cam.viewportWidth, cam.viewportHeight);
					ocam.position.set(cam.position);
					ocam.direction.set(cam.direction);
					ocam.near = cam.near;
					ocam.far = cam.far;
					cam = ocam;
					Gdx.app.log(TAG, "ORTHOGRAPHIC");
				}
			} else if (character == 'o') {
				cam.position.set(oPos);
				cam.direction.set(oDir);
			}
			return false;
		}
	}

	SpriteBatch batch;
	Texture texture;
	Sprite sprite;
	Camera cam;
	boolean adjXY;
	Vector3 startPos;
	Vector3 startDir;
	Vector3 oPos;
	Vector3 oDir;

	TextureAtlas textureAtlas;
	// MeshObj2 meshObj3;
	// MeshObj2 meshObj4;
	MapGrid mapGrid;

	@Override
	public void create() {
		ShaderProgram.pedantic = false;

		textureAtlas = new TextureAtlas(Gdx.files.internal("fell.pack"));
		{
			batch = new SpriteBatch();
			texture = new Texture("badlogic.jpg");
			sprite = new Sprite(texture);
			sprite.setSize(200, 200);
			sprite.setPosition(300, 100);
		}
		Gdx.app.log(TAG, "WINDOW SIZE=" + Gdx.graphics.getWidth() + ", "
				+ Gdx.graphics.getHeight());

		initMap();
		initCamera();

		Gdx.input.setInputProcessor(new MyInputAdapter());

		Gdx.app.log(TAG, "VP SIZE=" + cam.viewportWidth + ", "
				+ cam.viewportHeight);
	}

	void initMap() {
		int w = 5;
		int h = 5;

		mapGrid = new MapGrid();
		mapGrid.setPosition(50, 50);
		mapGrid.setSize(w, h, 80f);
		mapGrid.setSubdivide(4);
		mapGrid.setBaseZ(0);

		AtlasRegion grass = textureAtlas.findRegion("Grass01");
		for (int y = 0; y < h; y++) {
			for (int x = 0; x < w; x++) {
				mapGrid.setRegion(x, y, grass);
			}
		}
		mapGrid.build();

		GenZMap zmap = new GenZMap(mapGrid);
		zmap.setRandomSeed(0);
		zmap.addGenerator(new GenZConstant(50f));
		zmap.run();
	}

	void initMounds() {

	}

	void initCamera() {
		cam = new OrthographicCamera(Gdx.graphics.getWidth(),
				Gdx.graphics.getHeight());

		// cam.position.set(cam.viewportWidth / 2, cam.viewportHeight / 2,
		// cam.viewportWidth / 2);
		// cam.direction.set(0, 0, -1);

		startPos = new Vector3(cam.viewportWidth / 2, cam.viewportHeight / 2,
				cam.viewportWidth / 2);
		startDir = new Vector3(0, 0, -1);

		cam.position.set(startPos);
		cam.direction.set(startDir);

		cam.near = 1;
		cam.far = 1000;
		
		oPos = new Vector3(startPos.x, -300f, 570f);
		oDir = new Vector3(0, 0.92f, -1);
	}

	@Override
	public void render() {
		// Gdx.gl20.glViewport(0, 0, Gdx.graphics.getWidth(),
		// Gdx.graphics.getHeight());
		Gdx.gl20.glClearColor(0.6f, 0.2f, 0.2f, 1);
		Gdx.gl20.glClear(GL20.GL_COLOR_BUFFER_BIT);
		Gdx.gl20.glEnable(GL20.GL_TEXTURE_2D);
		Gdx.gl20.glEnable(GL20.GL_BLEND);
		Gdx.gl20.glBlendFunc(GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA);

		cam.update();

		// meshObj3.render(cam.combined);
		// meshObj4.render(cam.combined);

		// batch.setProjectionMatrix(cam.combined);
		// batch.begin();
		// sprite.draw(batch);
		// batch.end();

		mapGrid.render(cam.combined);
	}
}

File: Gdx/FellDungeon/app/src/main/java/com/dugsolutions/fell/FellDungeon.kt
package com.dugsolutions.fell

import com.badlogic.gdx.ApplicationAdapter
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.InputAdapter
import com.badlogic.gdx.graphics.*
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.glutils.ShaderProgram
import com.badlogic.gdx.math.Vector3
import com.dugsolutions.fell.data.MeshObj2
import com.dugsolutions.fell.db.DbCamera
import com.dugsolutions.fell.db.FellDatabaseHelper
import com.dugsolutions.fell.map.MapGrid
import com.dugsolutions.fell.map.gen2.GenZConstant
import com.dugsolutions.fell.map.gen2.GenZMap

class FellDungeon : ApplicationAdapter() {

    companion object {
        const val TAG = "FellDungeon"
        private const val ADJ = 0.01f
        private const val ZOOM = 0.02f
    }

    private inner class Box constructor(var cx: Int, var cy: Int, var radius: Int) {
        fun within(x: Int, y: Int): Boolean {
            val x0 = cx - radius
            val x1 = cx + radius
            val y0 = cy - radius
            val y1 = cy + radius
            return x in x0..x1 && y >= y0 && y <= y1
        }
    }

    private inner class MyInputAdapter : InputAdapter() {
        private var triggerX0MoveLeft: Int
        private var triggerX1MoveRight: Int
        private var triggerY0MoveUp: Int
        private var triggerY1MoveDown: Int
        private var centerTrigger: Box
        private var zoomIn = false
        fun move(screenX: Int, screenY: Int): Boolean {
            val camera = cam ?: return false
            if (screenX < triggerX0MoveLeft) {
                if (adjXY) {
                    camera.position.x += 10f
                } else {
                    camera.direction.x += ADJ
                }
            } else if (screenX > triggerX1MoveRight) {
                if (adjXY) {
                    camera.position.x -= 10f
                } else {
                    camera.direction.x -= ADJ
                }
            } else if (screenY < triggerY0MoveUp) {
                if (adjXY) {
                    camera.position.y -= 10f
                } else {
                    camera.direction.y -= ADJ
                }
            } else if (screenY > triggerY1MoveDown) {
                if (adjXY) {
                    camera.position.y += 10f
                } else {
                    camera.direction.y += ADJ
                }
            } else if (centerTrigger.within(screenX, screenY)) {
                if (camera is OrthographicCamera) {
                    if (zoomIn) {
                        camera.zoom += ZOOM
                    } else {
                        camera.zoom -= ZOOM

                    }
                }
            } else {
                return false
            }
            logCamera()
            return true
        }

        override fun touchDown(
            screenX: Int, screenY: Int, pointer: Int, button: Int
        ): Boolean {
            val camera = cam ?: return false
            if (button == 1) {
                camera.direction.set(startDir)
                camera.position.set(startPos)
                return true
            }
            return move(screenX, screenY)
        }

        override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
            return false
        }

        override fun touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean {
            return move(screenX, screenY)
        }

        override fun scrolled(amountX: Float, amountY: Float): Boolean {
            val camera = cam ?: return false
            camera.position.z += 10 * amountX
            log("Z=" + camera.position.z)
            return true
        }

        override fun keyTyped(character: Char): Boolean {
            val camera = cam ?: return false
            val manager = cameraManager ?: return false
            when (character) {
                'z' -> {
                    adjXY = !adjXY
                    if (adjXY) {
                        log("ADJUST POSITION")
                    } else {
                        log("ADJUST DIRECTION")
                    }
                }
                'c' -> {
                    if (camera is OrthographicCamera) {
                        val pcam = PerspectiveCamera(
                            45f,
                            camera.viewportWidth, camera.viewportHeight
                        )
                        pcam.position.set(camera.position)
                        pcam.direction.set(camera.direction)
                        pcam.near = camera.near
                        pcam.far = camera.far
                        cam = pcam
                        log("PERSPECTIVE")
                    } else {
                        val ocam = OrthographicCamera(
                            camera.viewportWidth, camera.viewportHeight
                        )
                        ocam.position.set(camera.position)
                        ocam.direction.set(camera.direction)
                        ocam.near = camera.near
                        ocam.far = camera.far
                        cam = ocam
                        log("ORTHOGRAPHIC")
                    }
                }
                'o' -> {
                    camera.position.set(oPos)
                    camera.direction.set(oDir)
                }
                's' -> {
                    manager.store()
                }
                'n' -> {
                    manager.load()
                }
                'd' -> {
                    manager.delete()
                }
                'Z' -> {
                    zoomIn = !zoomIn
                    (camera as? OrthographicCamera)?.zoom = 1f
                    logCamera()
                }
                'p' -> {
                    camera.position.x = 0f
                    camera.position.y = 0f
                    logCamera()
                }
            }
            return false
        }

        init {
            val width = Gdx.graphics.width
            val height = Gdx.graphics.height
            triggerX0MoveLeft = width / 4
            triggerX1MoveRight = width * 3 / 4
            triggerY0MoveUp = height / 4
            triggerY1MoveDown = height * 3 / 4
            centerTrigger = Box(width / 2, height / 2, width / 5)
        }
    }

    inner class CameraManager {
        private var list: ArrayList<DbCamera>? = null
        var index = -1

        fun load() {
            val camera = cam ?: return
            val list = list ?: return
            if (list.size > 0) {
                if (++index >= list.size) {
                    index = 0
                }
                list[index].upload(camera)
            }
        }

        fun store() {
            val camera = cam ?: return
            database?.save(camera)
        }

        fun delete() {
            val camera = cam ?: return
            val list = list ?: return
            if (list.size >= 0) {
                for (c in list) {
                    if (c.equals(camera)) {
                        database?.deleteCamera(c)
                        list.remove(c)
                    }
                }
            }
        }

        init {
            database?.let { db -> list = db.queryCameras() }
        }
    }

    private var batch: SpriteBatch? = null
    private var texture: Texture? = null
    private var sprite: Sprite? = null
    var cam: Camera? = null
    var adjXY = true
    var startPos: Vector3? = null
    var startDir: Vector3? = null
    var oPos: Vector3? = null
    var oDir: Vector3? = null
    private lateinit var textureAtlas: TextureAtlas

    private var mapGrid: MapGrid? = null
    var database: FellDatabaseHelper? = null
    var cameraManager: CameraManager? = null

    override fun create() {
        database = FellDatabaseHelper()
        database?.open()
        cameraManager = CameraManager()
        ShaderProgram.pedantic = false
        textureAtlas = TextureAtlas(Gdx.files.internal("fell.pack"))
        run {
            batch = SpriteBatch()
            texture = Texture("badlogic.jpg")
            val sprite = Sprite(texture)
            sprite.setSize(200f, 200f)
            sprite.setPosition(300f, 100f)
            this.sprite = sprite
        }
        Gdx.app.log(
            TAG, "WINDOW SIZE=" + Gdx.graphics.width + ", "
                    + Gdx.graphics.height
        )
        initMap()
        initCamera()
        Gdx.input.inputProcessor = MyInputAdapter()
        cam?.let { cam ->
            Gdx.app.log(
                TAG, "VP SIZE=" + cam.viewportWidth + ", " + cam.viewportHeight
            )
        }
    }

    private fun initMap() {
        val w = 5
        val h = 5
        val mapGrid = MapGrid()
        mapGrid.setPosition(0f, 0f)
        mapGrid.setSize(w, h, 80f)
        mapGrid.subdivide = 4
        mapGrid.baseZ = 0f
        val grass = textureAtlas.findRegion("Grass01")
        for (y in 0 until h) {
            for (x in 0 until w) {
                mapGrid.setRegion(x, y, grass)
            }
        }
        mapGrid.build()
        this.mapGrid = mapGrid
        val zmap = GenZMap(mapGrid)
        zmap.setRandomSeed(0)
        zmap.addGenerator(GenZConstant(50f))
        zmap.run()
        log("MAP SIZE = " + mapGrid.mapSize + ", pos=" + mapGrid.position)
    }

    private fun initCamera() {
        val cam = OrthographicCamera(
            Gdx.graphics.width.toFloat(),
            Gdx.graphics.height.toFloat()
        )
        this.cam = cam

        // cam.position.set(cam.viewportWidth / 2, cam.viewportHeight / 2,
        // cam.viewportWidth / 2);
        // cam.direction.set(0, 0, -1);
        val startPos = Vector3(
            cam.viewportWidth / 2, cam.viewportHeight / 2,
            cam.viewportWidth / 2
        )
        this.startPos = startPos
        startDir = Vector3(0f, 0f, -1f)
        cam.position.set(startPos)
        cam.direction.set(startDir)
        cam.near = 1f
        cam.far = 1000f
        oPos = Vector3(startPos.x, -300f, 570f)
        oDir = Vector3(0f, 0.92f, -1f)
    }

    override fun render() {
        // Gdx.gl20.glViewport(0, 0, Gdx.graphics.getWidth(),
        // Gdx.graphics.getHeight());
        val camera = cam ?: return
        Gdx.gl20.glClearColor(0.6f, 0.2f, 0.2f, 1f)
        Gdx.gl20.glClear(GL20.GL_COLOR_BUFFER_BIT)
        Gdx.gl20.glEnable(GL20.GL_TEXTURE_2D)
        Gdx.gl20.glEnable(GL20.GL_BLEND)
        Gdx.gl20.glBlendFunc(GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA)
        camera.update()

        batch?.let { batch ->
            sprite?.let { sprite ->
                batch.projectionMatrix = camera.combined
                batch.begin()
                sprite.draw(batch)
                batch.end()
            }
        }
        mapGrid?.render(camera.combined)
    }

    fun log(msg: String) {
        Gdx.app.log(TAG, msg)
    }

    private fun logCamera() {
        val camera = cam ?: return
        val sbuf = StringBuffer()
        sbuf.append("camera POS=" + camera.position.toString())
        if (camera.direction != Vector3(0f, 0f, -1f)) {
            sbuf.append(", DIR=" + camera.direction.toString())
        }
        if (camera is OrthographicCamera && camera.zoom != 1f) {
            sbuf.append(", ZOOM=" + camera.zoom)
        }
        log(sbuf.toString())
    }

}